### PR TITLE
[FLINK-15391][hive] DATE and TIMESTAMP partition columns don't work

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

Fix issues of read/write DATE and TIMESTAMP partition columns.

## Brief change log

  - When reading DATE/TIMESTAMP partition columns, make sure the value objects are converted to corresponding Flink objects.
  - When writing DATE/TIMESTAMP partition columns, make sure we convert to Hive objects before extracting the string values.
  - Add new tests

## Verifying this change

New test cases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
cc @JingsongLi @zjuwangg @bowenli86 @xuefuz

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		initConversions();
+		LinkedHashMap<String, String> partSpec = new LinkedHashMap<>();
+
+		for (int i = 0; i < partitionIndexes.length; i++) {
+			int index = partitionIndexes[i];
+			Object field = in.getField(index);
+			if (field != null) {
+				field = partColConversions[i].toHiveObject(field);
+			}
+			String partitionValue = field != null ? field.toString() : null;
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java]: minor: these can be consolidated on just `field != null`?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			if (partitionValue == null || "".equals(partitionValue)) {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java]: minor: StringUtils.isNullOrEmptyString(partitionValue)?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * A RowPartitionComputer that converts Flink objects to Hive objects before computing the partition value strings.
+ */
+public class HivePartitionComputer extends RowPartitionComputer {
+
+	private static final long serialVersionUID = 1L;
+
+	private final HiveShim hiveShim;
+	private final DataType[] columnTypes;
+	private HiveObjectConversion[] partColConversions;
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java]: Can we make `HiveObjectConversion` serializable?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 4]
```diff
@@ -286,7 +293,15 @@ private Object restorePartitionValueFromFromType(String valStr, DataType type) {
 			case DOUBLE:
 				return Double.valueOf(valStr);
 			case DATE:
-				return Date.valueOf(valStr);
+				return HiveInspectors.toFlinkObject(
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java]: You can  fix `AbstractOrcColumnVector.createHiveVectorFromConstant` too.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java (hunk 2 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java]: I don't think we have the `StringUtils.isNullOrEmptyString` util. We do have `StringUtils.isNullOrWhitespaceOnly` but it's not exactly the same as the logic here.

Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java]: `org.apache.commons.lang3.StringUtils.isEmpty` is OK.

**Comment by lirui-apache** [comment]
@bowenli86 @JingsongLi Thanks for the review. I rebased and addressed your comments.
I updated `ExpressionExtractor` to convert to Hive object when visiting a literal value. But since currently we don't support prune DATE/TIMESTAMP partition columns, we can't add test to verify that change.

**Comment by bowenli86** [comment]
LGTM, merging.

Regarding "we don't support prune DATE/TIMESTAMP partition columns, we can't add test to verify that change.", is that a bug or did we miss it in the first place? I think it should be part of the story of fully supporting partition pruning in HiveCatalog. Are we able to do it in 1.10?

**Pull request closed without merging**